backend/gs/file.go

```diff
@@ -25,84 +25,304 @@ const (
+			f.reader = f.tempFileWriter
+		} else {
+			// get object handle
+			h, err := f.getObjectHandle()
+			if err != nil {
+				return nil, err
+			}
+
+			// get range reader (from current cursor position to end of file)
+			reader, err := h.NewRangeReader(f.fileSystem.ctx, f.cursorPos, -1)
+			if err != nil {
+				return nil, err
+			}
+
+			// Set the reader to the body of the object
+			f.reader = reader
+		}
+	}
+	return f.reader, nil
+}
+
+// Seek implements the standard for io.Seeker.
 func (f *File) Seek(offset int64, whence int) (int64, error) {
-	if err := f.checkTempFile(); err != nil {
-		return 0, err
+	// get length of file
+	var length uint64
+	if f.writeCalled {
+		// if write has been called, then the length is the cursorPos
+		length = uint64(f.cursorPos)
+	} else {
+		var err error
+		length, err = f.Size()
+		if err != nil {
+			return 0, utils.WrapSeekError(err)
+		}
+	}
+
+	// invalidate reader (if any)
+	if f.reader != nil {
+		err := f.reader.Close()
+		if err != nil {
+			return 0, utils.WrapSeekError(err)
+		}
+
+		f.reader = nil
+	}
+
+	// invalidate s3Writer
+	if f.gcsWriter != nil {
+		// cancel s3Writer
+		f.cancelFunc()
+		f.cancelFunc = nil
+
+		f.gcsWriter = nil
+	}
+
+	// update seek position for tempFileWriter writer (if any)
+	if f.tempFileWriter != nil {
+		// seek tempFileWriter
+		_, err := f.tempFileWriter.Seek(offset, whence)
+		if err != nil {
+			return 0, utils.WrapSeekError(err)
+		}
 	}
-	return f.tempFile.Seek(offset, whence)
+
+	// update cursorPos
+	pos, err := utils.SeekTo(int64(length), f.cursorPos, offset, whence)
+	if err != nil {
+		return 0, utils.WrapSeekError(err)
+	}
+	f.cursorPos = pos
+
+	f.seekCalled = true
+	return f.cursorPos, nil
 }
 
-// Write implements the standard for io.Writer. A buffer is added to with each subsequent
-// write. Calling Close() will write the contents back to GCS.
-func (f *File) Write(data []byte) (n int, err error) {
-	if f.writeBuffer == nil {
-		// note, initializing with 'data' and returning len(data), nil
-		// causes issues with some Write usages, notably csv.Writer
-		// so we simply initialize with no bytes and call the buffer Write after
-		//
-		// f.writeBuffer = bytes.NewBuffer(data)
-		// return len(data), nil
-		//
-		// so now we do:
+// Write implements the standard for io.Writer.  Note that writes are not committed to GCS until CLose() is called.
+func (f *File) Write(data []byte) (int, error) {
+	// check/initialize for writer
+	err := f.initWriters()
+	if err != nil {
+		return 0, utils.WrapWriteError(err)
+	}
+
+	// write to tempfile
+	written, err := f.tempFileWriter.Write(data)
+	if err != nil {
+		return 0, utils.WrapWriteError(err)
+	}
 
-		f.writeBuffer = bytes.NewBuffer([]byte{})
+	// write to gcs
+	if f.gcsWriter != nil {
+		// write to s3
+		gcsWritten, err := f.gcsWriter.Write(data)
+		if err != nil {
+			return 0, utils.WrapWriteError(err)
+		}
 
+		// ensure both writes are the same
+		if written != gcsWritten {
```

Comment:
I might be missing the reasoning behind the dual-writes. Mind giving a brief description on why that's necessary?

Reply:
Updated comments in https://github.com/C2FO/vfs/pull/173/commits/ff76f1ccde3c5b5321677b20d6bb83373001894a